<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>

# Backpropagation Practice

## *Data Science Unit 4 Sprint 2 Assignment 2*

Implement a 3 input, 4 node hidden-layer, 1 output node Multilayer Perceptron on the following dataset:

| x1 | x2 | x3 | y |
|----|----|----|---|
| 0  | 0  | 1  | 0 |
| 0  | 1  | 1  | 1 |
| 1  | 0  | 1  | 1 |
| 0  | 1  | 0  | 1 |
| 1  | 0  | 0  | 1 |
| 1  | 1  | 1  | 0 |
| 0  | 0  | 0  | 0 |

If you look at the data you'll notice that the first two columns behave like an XOR gate while the last column is mostly just noise. Remember that creating an XOR gate was what the perceptron was criticized for not being able to learn. 

In [27]:
import numpy as np
X = np.array(([0,0,1],
              [0,1,1],
              [1,0,1],
              [0,1,0],
              [1,0,0],
              [1,1,1],
              [0,0,0]), dtype=float)

# Exam Scores
y = np.array(([0],
              [1],
              [1],
              [1],
              [1],
              [0],
              [0]), dtype=float)

In [95]:
class NeuralNetworkMultiPerceptron:
    def __init__(self):
        #initalize input layer, hidden layer and output layer
        self.numInputLayerSize = 3
        self.numHiddenLayerSize =4
        self.numOutputLayerSize =1 
        #initalize weights with random number to start out with
        self.weight1 = np.random.rand(self.numInputLayerSize, self.numHiddenLayerSize)
        self.weight2 = np.random.rand(self.numHiddenLayerSize, self.numOutputLayerSize)    
    def sigmoid(self, s):
        return 1/(1+np.exp(-s))
    def DerivateSigmoid(self, s):
        return (np.exp(-s) / (1+np.exp(-s))**2)
    def ShootThisForward(self, X):
        self.DoTOfInputWights = np.dot(X, self.weight1)
        self.InsideHiddenLayerOne = self.sigmoid(self.DoTOfInputWights)
        self.SumOfHiddenLayer = np.dot(self.InsideHiddenLayerOne, self.weight2)
        SigmoidOfHiddenLayer = self.sigmoid(self.SumOfHiddenLayer)
        return SigmoidOfHiddenLayer
    def letgoBackWord(self, X, Real_Output, outputOfAlgorith):
        self.error_output = Real_Output - self.DerivateSigmoid(outputOfAlgorith)
        #let unsquish these stuff to sigmoid f(x), but we will be using the error. So let find out the angle of sigmoid
        #between the error and the giving output to the original output.
        self.adjustedTheVector_OutTo_H = self.error_output * self.DerivateSigmoid(outputOfAlgorith)
        self.error_hiddenTwo = self.adjustedTheVector_OutTo_H.dot(self.weight2.T)
        self.adjusted_hiddenVector_HtoIn = self.error_hiddenTwo * self.DerivateSigmoid(self.InsideHiddenLayerOne)
        #
        self.weight1 += X.T.dot(self.adjusted_hiddenVector_HtoIn)
        
        self.weight2 += self.InsideHiddenLayerOne.T.dot(self.adjustedTheVector_OutTo_H)
    def train(self, X, Real_Output):
        outputOfAlgorith=self.ShootThisForward(X)
        self.letgoBackWord(X,Real_Output, outputOfAlgorith)

In [92]:
    
##### Your Code Here #####
nn = NeuralNetworkMultiPerceptron()
help(nn)

Help on NeuralNetworkMultiPerceptron in module __main__ object:

class NeuralNetworkMultiPerceptron(builtins.object)
 |  Methods defined here:
 |  
 |  DerivateSigmoid(self, s)
 |  
 |  ShootThisForward(self, X)
 |  
 |  __init__(self)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  letgoBackWord(self, X, Real_Output, outputOfAlgorith)
 |  
 |  sigmoid(self, s)
 |  
 |  train(self, X, Real_Output)
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors defined here:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)



In [93]:
nn.train(X,y)

ValueError: shapes (7,3) and (60000,784) not aligned: 3 (dim 1) != 60000 (dim 0)

In [96]:
# Number of Epochs / Iterations
for i in range(10000):
    if (i+1 in [1,2,3,4,5]) or ((i+1) % 1000 ==0):
        print('+' + '---' * 3 + f'EPOCH {i+1}' + '---'*3 + '+')
        print('Input: \n', X)
        print('Actual Output: \n', y)
        print('Predicted Output: \n', str(nn.ShootThisForward(X)))
        print("Loss: \n", str(np.mean(np.square(y - nn.letgoBackWord(X)))))
    nn.train(X,y)

+---------EPOCH 1---------+
Input: 
 [[0. 0. 1.]
 [0. 1. 1.]
 [1. 0. 1.]
 [0. 1. 0.]
 [1. 0. 0.]
 [1. 1. 1.]
 [0. 0. 0.]]
Actual Output: 
 [[0.]
 [1.]
 [1.]
 [1.]
 [1.]
 [0.]
 [0.]]


ValueError: shapes (7,3) and (60000,784) not aligned: 3 (dim 1) != 60000 (dim 0)

## Try building/training a more complex MLP on a bigger dataset.

Use the [MNIST dataset](http://yann.lecun.com/exdb/mnist/) to build the cannonical handwriting digit recognizer and see what kind of accuracy you can achieve. 

If you need inspiration, the internet is chalk-full of tutorials, but I want you to see how far you can get on your own first. I've linked to the original MNIST dataset above but it will probably be easier to download data through a neural network library. If you reference outside resources make sure you understand every line of code that you're using from other sources, and share with your fellow students helpful resources that you find.


### Parts
1. Gathering & Transforming the Data
2. Making MNIST a Binary Problem
3. Estimating your Neural Network (the part you focus on)

### Gathering the Data 

`keras` has a handy method to pull the mnist dataset for you. You'll notice that each observation is a 28x28 arrary which represents an image. Although most Neural Network frameworks can handle higher dimensional data, that is more overhead than necessary for us. We need to flatten the image to one long row which will be 784 values (28X28). Basically, you will be appending each row to one another to make on really long row. 

In [98]:
import numpy as np
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

In [70]:
# input image dimensions
img_rows, img_cols = 28, 28

In [71]:
# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [72]:
y_train

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [73]:
x_train = x_train.reshape(x_train.shape[0], img_rows * img_cols)
x_test = x_test.reshape(x_test.shape[0], img_rows * img_cols)

# Normalize Our Data
x_train = x_train / 255
x_test = x_test / 255

In [74]:
# Now the data should be in a format you're more familiar with
x_train.shape

(60000, 784)

### Making MNIST a Binary Problem 
MNIST is multiclass classification problem; however we haven't covered all the necessary techniques to handle this yet. You would need to one-hot encode the target, use a different loss metric, and use softmax activations for the last layer. This is all stuff we'll cover later this week, but let us simply the problem for now: Zero or all else.

In [75]:
import numpy as np

y_temp = np.zeros(y_train.shape)
y_temp[np.where(y_train == 0.0)[0]] = 1
y_train = y_temp

y_temp = np.zeros(y_test.shape)
y_temp[np.where(y_test == 0.0)[0]] = 1
y_test = y_temp

In [76]:
# A Nice Binary target for ya to work with
y_train

array([0., 1., 0., ..., 0., 0., 0.])

In [100]:
train_y_one_hot = to_categorical(y_train)
test_y_one_hot = to_categorical(y_test)
### Estimating Your `net

In [115]:
from keras.layers import Conv2D, MaxPooling2D, Dense, Flatten
from keras.models import Sequential
import numpy as np
#import mnist
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Dense, Flatten
from keras.utils import to_categorical
# input image dimensions
img_rows, img_cols = 28, 28

AttributeError: module 'tensorflow._api.v1.keras.datasets.mnist' has no attribute 'train_images'

In [ ]:
# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [ ]:
y_train

In [117]:
'''Trains a simple convnet on the MNIST dataset.
Gets to 99.25% test accuracy after 12 epochs
(there is still a lot of margin for parameter tuning).
16 seconds per epoch on a GRID K520 GPU.
'''

from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K

batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Instructions for updating:
Use tf.cast instead.
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 107s 2ms/step - loss: 0.2768 - acc: 0.9163 - val_loss: 0.0586 - val_acc: 0.9816
Epoch 2/12
60000/60000 [==============================] - 107s 2ms/step - loss: 0.0925 - acc: 0.9723 - val_loss: 0.0451 - val_acc: 0.9849
Epoch 3/12
60000/60000 [==============================] - 108s 2ms/step - loss: 0.0688 - acc: 0.9800 - val_loss: 0.0360 - val_acc: 0.9884
Epoch 4/12
60000/60000 [==============================] - 107s 2ms/step - loss: 0.0582 - acc: 0.9836 - val_loss: 0.0342 - val_acc: 0.9889
Epoch 5/12
60000/60000 [==============================] - 106s 2ms/step - loss: 0.0488 - acc: 0.9851 - val_loss: 0.0317 - val_acc: 0.9899
Epoch 6/12
60000/60000 [

In [ ]:
model.fit(x_train, train_y_one_hot, batch_size=64, epochs=10)

## Stretch Goals: 

- Make MNIST a multiclass problem using cross entropy & soft-max
- Implement Cross Validation model evaluation on your MNIST implementation 
- Research different [Gradient Descent Based Optimizers](https://keras.io/optimizers/)
 - [Siraj Raval the evolution of gradient descent](https://www.youtube.com/watch?v=nhqo0u1a6fw)
- Build a housing price estimation model using a neural network. How does its accuracy compare with the regression models that we fit earlier on in class?